pkg/server/controllers.go

```diff
@@ -849,6 +850,39 @@ func (s *Server) installAPIExportController(ctx context.Context, config *rest.Co
 	})
 }
 
+func (s *Server) installAPIExportEndpointSliceController(ctx context.Context, config *rest.Config, server *genericapiserver.GenericAPIServer) error {
+	config = rest.CopyConfig(config)
+	config = rest.AddUserAgent(config, apiexportendpointslice.ControllerName)
+
+	kcpClusterClient, err := kcpclientset.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	c, err := apiexportendpointslice.NewController(
+		s.KcpSharedInformerFactory.Apis().V1alpha1().APIExportEndpointSlices(),
+		// ClusterWorkspaceShards and APIExports get retrieved from cache server
```

Comment:
Does a shard have some local storage for its' own definition? Surely also there will be APIExports on the shard, that do not need to be requested from the cache - we can progress in processing that data even when the cache is down, right?

Reply:
> Does a shard have some local storage for its' own definition? 

Currently not, it is all in the root workspace of the root shard. As per the above I think we would need it.

> here will be APIExports on the shard, that do not need to be requested from the cache - we can progress in processing that data even when the cache is down, right?

That's true but I am not sure it is worth it. We need the shard information for getting more than a single endpoint. This can only be retrieved from the cache server. I would rather keep things "frozen" when the cache server cannot be reached: existing services are not impacted by the unavailability of the cache server as long as the shards don't change. Looking only at the information on the local shard, which is currently only available for the root shard if the APIExport has been created there would only give a partial picture that may be worse than the stale snapshot captured in the APIExportEndpointSlices before the unavailability of the cache server.

Reply:
I think that is against our principle of single-shard functionality without external data. We need to support "working as much as possible" when the larger system has network partitions.

Reply:
@p0lyn0mial would appreciate your thoughts on this

Reply:
> Does a shard have some local storage for its' own definition? Surely also there will be APIExports on the shard, that do not need to be requested from the cache - we can progress in processing that data even when the cache is down, right?

yes, so, as of today our approach is to use two informers (maybe in the future we could unify and provide a single instance/interface).

The approeach so far was to first check the local informer and falling back to the second backed by the cache server. For example:
```
getAPIExport: func(clusterName logicalcluster.Name, name string) (*apisv1alpha1.APIExport, error) {
  apiExport, err := localAPIExportInformer.Lister().Cluster(clusterName).Get(name)
  if errors.IsNotFound(err) {
    return cacheApiExportInformer.Lister().Cluster(clusterName).Get(name)
  }
  return apiExport, err
}
```

I think we could apply this strategy to all places that require working with the caching layer.

When it comes to working with `Shard` resources in an event of a failure, we could at least register the shard we are currently running on since all we need is `Spec.VirtualWorkspaceURL`.

Does it make sense?

Reply:
> I think we could apply this strategy to all places that require working with the caching layer.

I think that we need to discriminate between two types of requests:
- I am looking for APIExport A, or another global resource type, and the existing approach: check locally and then the cache looks ok to me.
- I want to build a complete view and I am looking for **all** shards. This is the scenario here. In that case I don't think that local requests bring some added value. 

> When it comes to working with Shard resources in an event of a failure, we could at least register the shard we are currently running on since all we need is Spec.VirtualWorkspaceURL

If the APIExportEndpointSlice was just created, nothing is broken, just not yet working. If the cache server is "temporarily" not accessible it is debatable but my gut feeling is that it is simpler, clearer, cleaner to wait for it to get reachable again and the reconciliation to happen.
If the APIExportEndpointSlice already existed, and had already been reconciled once, looking at the local shard only, may (not today) return a single shard. Reconciliation against that may mean to remove hundred of valid existing URLs and possibly one invalid (if a shard change was the trigger for the reconciliation) and replace them by a single URL (the local shard). I would argue that it is far worse than doing nothing, keeping  the stale state and waiting for the cache server to be reachable again to have the state finally up-to-date.

to come back to:

> Does a shard have some local storage for its' own definition?

As per my previsous answer: No. I am proposing to change that, to stop having the root shard being special and the cache server becoming **the** preferred communication mechanism between shards. I feel however that it is orthogonal to this pull request.

Reply:
Added a commit following our [slack discussion](https://kubernetes.slack.com/archives/C021U8WSAFK/p1672668119924479). My preference would be without it.